README.md
# DeLuciatoR

A few tools to help the DeLucia lab make R plots look the way Evan wants them to look. Improvements from DeLucians are encouraged, improvements from anyone else are welcome but subject to judgement calls on whether they fit the lab aesthetic.

Basic goals of "look the way Evan wants", not all of which are implemented yet:

- [ ] All text and markings large enough to be *easily* readable when reduced to a tiny subfigure or crammed into a low-rez powerpoint slide. Mostly implemented, needs your feedback on which sizes need refinement.
- [x] 1.33:1 aspect ratio (Y axis 3/4 as tall as X axis is long).
- [x] Axis ticks facing inward.
- [x] Axis ticks on all four sides. Requires package 'ggplotTicks' (see below).
- [x] Y-axis numbers set horizontally.
- [ ] Colors, if any, should be interpretable by dichromats and by those who printed the paper in black and white. Not yet implemented, but the [viridis](https://github.com/sjmgarnier/viridis) package may help.
- [ ] It should be as easy as possible to rescale one plot to look good at any size you need it. Partly implemented: Most axis and legend elements scale nicely, but you still need to manually  adjust the size of plotting symbols and annotation text.

The package currently provides:

* One ggplot2 theme, mostly optimized for X-Y plots but probably suitable for most scatter/line plots and a good starting point for most bar/box/density plots.
* Some functions to save correctly-shaped images of fixed-aspect-ratio plots: Specify a maximum height and width, get the biggest image that fits inside them. Compare this to the default pdf and png graphics functions, which make you specify an exact height & width (tedious to calculate for multipanel plots) and add oodles of whitespace if you get the ratio wrong.

Please help me add to the toolbox! Let's add more shortcuts for common ggplot actions, and some themes for Lattice plots... and for plain old base graphics too!


## Installation

The easy way, in an R session:

```{r}
install.packages("devtools")
devtools::install_github("infotroph/DeLuciatoR")
library("DeLuciatoR")
```

The hard way, in a terminal window (you only need to do this if you want to make your own changes to the package):

```{r}
git clone https://github.com/infotroph/DeLuciatoR.git
# Edit any files you like here
R CMD build DeLuciatoR
R CMD check DeLuciator_0.1.0.tar.gz
# If R complains, go back and fix any errors here
R CMD install DeLuciator_0.1.0.tar.gz
```


## Setting the theme

Use `theme_ggEHD` the same way you use the built-in ggplot themes: Add it to the original plotting call, or save the object and modify the theme later.

```{r}
p = (ggplot(iris, aes(Sepal.Length, Petal.Width, shape=Species))
	+ geom_point(size=2)
	+ geom_smooth(method="lm")
	+theme_ggEHD())

plot(p) # "...hmm, that text is too small..."
plot(p + theme_ggEHD(base_size=24)) # "...better, but the legend needs to move..."
p = p + theme_ggEHD(base_size=24) + theme(legend.position=c(0.85, 0.25)) # "...Aaah, good."
```

![Iris dataset styled three ways](demo_images/iris.png)

## Putting ticks on all sides

The ggplot2 package provides no built-in way to put ticks on all sides of the plot. To get around this, I wrote [an R package](https://github.com/infotroph/ggplotTicks) that mirrors the axes after the rest of the plot is constructed. I will eventually make this package use it automatically; for now, grab it from GitHub and wrap your plots in a call to `mirror_ticks`:

```{r}
install_github("infotroph/ggplotTicks")
library("ggplotTicks")

p_two = (ggplot(CO2, aes(conc, uptake, color=Treatment))
	+ geom_point()
	+ facet_wrap(~Type)
	+ theme_ggEHD())

p_four = mirror_ticks(p_two)
```
`p_four` is now a modified ggplot, with plot ticks on all four sides. It behaves *almost* exactly like a ggplot, *except* that adding or removing facets will remove the mirroring:

```{r}
# has ticks on all sides
plot(p_four)

# has ticks on all sides
plot(p_four
	+ coord_trans(x="log", y="log")
	+ xlab("Ceci n'est pas une linear axis"))

# loses mirroring; identical to p_two + facet_wrap(~)
plot(p_four + facet_wrap(~Plant))

# has ticks on all sides
plot(mirror_ticks(p_four + facet_wrap(~Plant)))
```

![CO2 dataset plotted four ways](demo_images/co2.png)

See the `ggplotTicks` documentation for more details.


## Image size calculations

Once you have a plot you like, save it. DeLuciatoR makes that easy:

```{r}
ggsave_fitmax(
	filename="iris_plot.pdf",
	plot=p,
	maxwidth=6.5,
	maxheight=9)
```

Why is this better than `ggsave` or `pdf(...); plot(...); dev.off()` or `File > Save As...`? In short, because `ggsave_fitmax` is *simple*, *consistent* no matter whether called from a script or an interactive session, and *repeatable*.

When you save one plot during an interactive session, you generally save the exact image you see in front of you. Need to remake it in a different color? Hope you can drag your new plot window back to *exactly* this height and width!

Plotting from a script makes it easy to reproduce a plot exactly, but standard graphics functions such as `pdf()` need you to specify the exact dimensions of the image, which are not trivial to calculate: "OK, I want it to be 16 cm wide and it has two panels side-by side, each with an aspect ratio of 0.75, and the left axis uses about 2 cm, so that's `7 * 3/4` is 5.25 cm high, but the bottom axis will add some—*right*, the legend takes up space too, so that pushed the panel width down to—*Aaaaagh*!"

`ggsave_fitmax` knows how to do all these size calculations for you, without screaming. Simply specify the largest allowable height and width, and it will save a correctly-shaped image, with no stretching or extra whitespace, that fits inside the limits you specify.


## Prior Art

The basic idea behind `DeLuciatoR` is to provide some default graphical settings that align with one particular person's (Evan DeLucia's) opinions on how to produce scientific graphics. For more ideas on what's possible from opinionated personal packages built on top of ggplot, see:

* Claus Wilke's [cowplot](https://cran.r-project.org/web/packages/cowplot/vignettes/introduction.html)
* Auriel Fournier's [rel](https://github.com/aurielfournier/rel), especially for [theme_krementz](https://github.com/aurielfournier/rel/blob/master/R/theme_krementz.R)
* Jeffrey Arnold's [ggthemes](https://github.com/jrnold/ggthemes), an ever-growing collection of ggplot themes and color scales
* Robert Wilson's [ggplot2-theme](https://github.com/robertwilson190/ggplot2-theme)
* Noam Ross's [noamtools](https://github.com/noamross/noamtools)
* Jenny Bryan's [ggplot tutorial](https://github.com/jennybc/ggplot2-tutorial)


NEWS
v. 0.1.2, 2016-12-07:
	* Text now explicitly set to black instead of inheriting whatever theme_bw gives it (#12)
	* Remove boxes around legend keys (#11)
v. 0.1.1, 2016-05-23:
	* bug fix: get_dims no longer fails on plots with null units inside more complex unit objects (e.g. multipanel ggplots).
v. 0.1.0, 2016-03-04:
	* get_dims should now respect aspect ratios set in global theme (#3).
	* new function `ggsave_fitmax` to save fixed-aspect plots without extra whitespace.
	* `png_ggsized` is deprecated. Use `ggsave_fitmax` instead.
	* Argument changes in `theme_ggEHD`:
		- Gains argument `base_family`, defaulting to "".
		- Loses `...`, which only worked for elements not set by the theme.
		- This means theme_ggEHD now takes the same arguments as the built-in ggplot themes.
		- If you want to pass additional arguments in `...`, use `theme_ggEHD()+theme(...)`.
v. 0.0.2, 2016-01-20:
	* Now compatible with ggplot2 >= 2.0 (1.0.x *should* still work, at least for now).
v. 0.0.1, 2015-10-01: 
	* Restructured previously existing code as an R package.
 	* Added new functions `get_dims` and `png_ggsized`.

tests/example.R
library(ggplot2)
library(DeLuciatoR)
library(gridExtra) # for multiplot demo
library(ggplotTicks)

# An example scatterplot from the R built-in `mtcars` dataset:
# Plot weight converted to kg against miles per US gallon
carplot = (ggplot(mtcars, aes(wt*453.592, mpg))
	+geom_smooth(method="lm", se=TRUE, linetype="dotted", color="black")
	+geom_point()
	+xlab("Vehicle weight (kg)")
	+ylab("MPG"))

# Add axis ticks to the top and right sides
carplot = mirror_ticks(carplot)

# Now let's see some themes.
# First something you might use in a small on-screen display...	
car_small = carplot + theme_ggEHD(12)

# ...then one with room to blow it up.
# Note that themes do not affect point size, so I set those manually.
car_big = carplot + geom_point(size=4) + theme_ggEHD(32)

# Let's put both plots side-by-side to see the differences.
# Try resizing the plot window and watch what happens.
# The settings you want depend on your image size!
grid.arrange(car_small, car_big, nrow=1)


# README examples
p = (ggplot(iris, aes(Sepal.Length, Petal.Width, shape=Species))
	+ geom_point(size=2)
	+ geom_smooth(method="lm")
	+theme_ggEHD())

plot(p) # "...hmm, that text is too small..."
plot(p + theme_ggEHD(base_size=24)) # "...better, but the legend needs to move..."
p = p + theme_ggEHD(base_size=24) + theme(legend.position=c(0.85, 0.25)) # "...Aaah, good."

ggsave_fitmax(
	filename="iris_plot.pdf",
	plot=p,
	maxwidth=6.5,
	maxheight=9)



R/get_dims.R
# Given a plot with fixed *panel* aspect ratios, find dimensions that
# fit the *whole plot* inside a given maximum height and width.

get_dims = function(ggobj, maxheight, maxwidth=maxheight, units="in", ...){

	# Check for common cases where plot ratios are unconstrained
	# ==> don't need to build anything, can infer that the plot will
	# fill the whole area.
	if(inherits(ggobj, "ggplot")
			&& !isTRUE(ggobj$respect)
			&& is.null(ggobj$theme$aspect.ratio)
			&& is.null(ggobj$coordinates$ratio)
			&& is.null(theme_get()$aspect.ratio)){
		return(list(height=maxheight, width=maxwidth))
	}

	# Open a temporary plotting device to do the calculations inside.
	# This is for two reasons: Ultimately because unit conversions
	# are device-dependent, but proximally because otherwise Grid will
	# open one for us and leave it open for the user to deal with.
	tmpf = tempfile(pattern="dispos-a-plot", fileext= ".png")
	png(
		filename=tmpf,
		height=maxheight,
		width=maxwidth,
		units=units,
		res=120,
		...)
	on.exit({dev.off(); unlink(tmpf)})

	# Supported plot types: Supposedly handles anything ggplot can produce.
	# Probably works on other gtables, e.g. gridExtra::arrangeGrob,
	# but these are less tested. TODO: Lattice output?
	if(inherits(ggobj, "ggplot")){
		g = ggplotGrob(ggobj)
	}else if (inherits(ggobj, "gtable")){
		g = ggobj
	}else{
		stop("Don't know how to get sizes for object of class ",
			deparse(class(ggobj)))
	}

	# Our approach relies on the quirk that
	# grid::convertUnit treats null units as 0.
	# If this changes, it will be rewrite time.
	stopifnot(grid::convertUnit(unit(1, "null"), "in", "x", valueOnly=T) == 0)

	# This sum gives the dimensions filled by *fixed-size* grobs.
	# We'll divide the remaining available space between rows/columns
	# in proportion to their size in null units.
	known_ht = sum(grid::convertHeight(g$heights, units, valueOnly=TRUE))
	known_wd = sum(grid::convertWidth(g$widths, units, valueOnly=TRUE))

	free_ht = maxheight - known_ht
	free_wd = maxwidth - known_wd

	# Find rows & columns specified in null units.
	# This is a convoluted process because unit names are potentially many layers deep in
	# unit.arithmetic or unit.list objects. Rather than access them directly, we'll compute
	# fixed dimensions twice, once as normal and once after replacing all the null units
	# with inches. Then the difference between these, even though reported as inches,
	# is the dimension in nulls.
	all_null_rowhts = (grid::convertHeight(.null_as_if_inch(g$heights), "in", valueOnly=TRUE)
		- grid::convertHeight(g$heights, "in", valueOnly=TRUE))
	all_null_colwds = (grid::convertWidth(.null_as_if_inch(g$widths), "in", valueOnly=TRUE)
		- grid::convertWidth(g$widths, "in", valueOnly=TRUE))
	null_rowhts = all_null_rowhts[all_null_rowhts > 0]
	null_colwds = all_null_colwds[all_null_colwds > 0]

	panel_asps = matrix(null_rowhts, ncol=1) %*% matrix(1/null_colwds, nrow=1)

	# Handle cases where height or width is fully determined. 
	# TODO: Incomplete! As written, these cases have a zero-element panel_asps
	# that throws "subscript out of bounds" in the *_if_max* calculations.
	# Need to understand how aspect ratios should (/actually do) work
	# when only one side is in null units. Does g$respect change behavior?
	# 
	# Maybe: If height fixed and width has nulls, return list(min(maxheight, known_ht), max_colwds)?
	# if(length(null_rowhts) == 0){
	# 	# All rows have fixed height, compute
	# 	maxheight = min(maxheight, known_ht)
	# 	free_ht = 0
	# 	max_rowhts = g$heights
	# }
	# if(length(null_rowhts) == 0){
	# 	# All cols have fixed width
	# 	maxwidth = min(maxwidth, known_wd)
	# 	free_wd = 0
	# 	max_colwds = g$widths
	# }

	max_rowhts = free_ht / sum(null_rowhts) * null_rowhts
	max_colwds = free_wd / sum(null_colwds) * null_colwds
	# print(c("max_rowhts: ", max_rowhts, " max_colwds: ", max_colwds))
		
	rowhts_if_maxwd = max_colwds[1] * panel_asps[,1]
	colwds_if_maxht = max_rowhts[1] / panel_asps[1,]
	# print(c("rowhts_if_maxwd: ", rowhts_if_maxwd, " colwds_if_maxht: ", colwds_if_maxht))

	height = min(maxheight, known_ht + sum(rowhts_if_maxwd))
	width = min(maxwidth, known_wd + sum(colwds_if_maxht))

	return(list(height=height, width=width))
}

# Internal helper function:
# Treat all null units in a unit object as if they were inches.
# This is a bad idea in gneral, but I use it here as a workaround.
# Extracting unit names from non-atomic unit objects is a pain,
# so questions like "which rows of this table layout have null heights?"
# are hard to answer. To work around it, I exploit an (undocumented!)
# quirk: When calculating the size of a table layout inside a Grid plot,
# convertUnit(...) treats null units as zero.
# Therefore
#	(convertHeight(grob_height, "in", valueOnly=TRUE)
#	- convertHeight(null_as_if_inch(grob_height), "in", valueOnly=TRUE))
# does the inverse of convertUnit: It gives the sum of all *null* heights
# in the object, treating *fixed* units as zero.
#
# Warning: I repeat, this approach ONLY makes any sense if
#	convertUnit(unit(1, "null"), "in", "x", valueOnly=T) == 0
# is true. Please check that it is before calling this code.
.null_as_if_inch = function(u){
	if(!grid::is.unit(u)) return(u)
	if(is.atomic(u)){
		if("null" %in% attr(u, "unit")){
			d = attr(u, "data")
			u = unit(
				x=as.vector(u),
				units=gsub("null", "in", attr(u, "unit")),
				data=d)
		}
		return(u)
	}
	if(inherits(u, "unit.arithmetic")){
		l = .null_as_if_inch(u$arg1)
		r = .null_as_if_inch(u$arg2)
		if(is.null(r)){
			args=list(l)
		}else{
			args=list(l,r)
		}
		return(do.call(u$fname, args))
	}
	if(inherits(u, "unit.list")){
		return(do.call(grid::unit.c, lapply(u, .null_as_if_inch)))
	}
	return(u)
}


R/theme_ggEHD.R
theme_ggEHD = function(base_size=18, base_family=""){
	thm = (theme_bw(base_size=base_size, base_family=base_family) %+% theme(
		panel.grid.major = element_blank(),
		panel.grid.minor = element_blank(),
		rect = element_rect(size=rel(0.75)),
		line = element_line(size=rel(1)),
		axis.line = element_line(size=rel(1)),
		axis.ticks = element_line(size=rel(1)),
		axis.ticks.length = unit(-(base_size*0.5), "points"),
		plot.margin = unit(base_size*c(1,1,1,1), "points"),
		text=element_text( # Can we inherit some of these?
			family=base_family,
			face="plain",
			size=base_size,
			color="black",
			hjust=0.5,
			vjust=0.5,
			angle=0),
		axis.text=element_text(size=rel(0.8), color = "black"),
		axis.title.x=element_text(vjust=-1),
		axis.title.y=element_text(vjust=2),
		legend.key=element_blank(),
		aspect.ratio=0.75
	))

	if(packageVersion("ggplot2") >= 2.0){
		return(thm %+% theme(
			axis.text.x = element_text(margin=margin(t=base_size*1.5, unit="points")),
			axis.text.y = element_text(margin=margin(r=base_size*1.5, unit="points"))))
	}else{
		return(thm %+% theme(
			axis.ticks.margin = unit((base_size*1.5), "points")))
	}
}


example.r
library(ggplot2)
library(gridExtra) # for multiplot demo
library(ggplotTicks)

# An example scatterplot from the R built-in `mtcars` dataset:
# Plot weight converted to kg against miles per US gallon
carplot = (ggplot(mtcars, aes(wt*453.592, mpg))
	+geom_smooth(method="lm", se=TRUE, linetype="dotted", color="black")
	+geom_point()
	+xlab("Vehicle weight (kg)")
	+ylab("MPG"))

# Add axis ticks to the top and right sides
carplot = mirror_ticks(carplot)

# Now let's see some themes.
# First something you might use in a small on-screen display...	
car_small = carplot + theme_ggEHD(12)

# ...then one with room to blow it up.
# Note that themes do not affect point size, so I set those manually.
car_big = carplot + geom_point(size=4) + theme_ggEHD(32)

# Let's put both plots side-by-side to see the differences.
# Try resizing the plot window and watch what happens.
# The settings you want depend on your image size!
grid.arrange(car_small, car_big, nrow=1)


# README examples
p = (ggplot(iris, aes(Sepal.Length, Petal.Width, shape=Species))
	+ geom_point(size=2)
	+ geom_smooth(method="lm")
	+theme_ggEHD())

plot(p) # "...hmm, that text is too small..."
plot(p + theme_ggEHD(base_size=24)) # "...better, but the legend needs to move..."
p = p + theme_ggEHD(base_size=24) + theme(legend.position=c(0.85, 0.25)) # "...Aaah, good."

ggsave_fitmax(
	filename="iris_plot.pdf",
	plot=p,
	maxwidth=6.5,
	maxheight=9)

